docs/source/ops/monitoring.rst

```diff
@@ -0,0 +1,130 @@
+
+Google's *Monitoring Distributed Systems* book about what they call the `Four Golden Signals <https://sre.google/sre-book/monitoring-distributed-systems/#xref_monitoring_golden-signals>`_ has a great explanation and definition of useful metrics:
+
+Latency
+  Requests, but also errors take time, so don't discard them.
+
+Traffic
+  What constitutes "Traffic" depends on the nature of your system.
+
+Errors
+  (The *rate* of ) failed requests.
+
+Saturation
+  How "full" your service is.  Take action (i.e. page a human) before service degrades.
+
+"If you measure all four golden signals and page a human when one signal is problematic (or, in the case of saturation, nearly problematic), your service will be at least decently covered by monitoring."
+
+
+RED method for services ("request-scoped", "external view")
+```````````````````````````````````````````````````````````
+
+Request rate, Errors, Duration (+ Saturation?)
+
+* Instrument everything that takes time and could fail
+   * "In contrast to logging, services should instrument every meaningful number available for capture." (Peter Bourgon)
+
+* Plot 99th percentile, 50th percentile and average
+   * 50th and average should be close - else something is wrong
+   * Averages sum neatly - Service latency average should be sum of child service latencies
+
+
+USE method for resources ("resource-scoped", "internal view")
+`````````````````````````````````````````````````````````````
+
+Utilization, Saturation, Errors:
+
+* CPUs
+   * Saturation (Per core? So we catch maxxed out single cores? Would that work?)
+* Memory
+   * Capacity
+* Network
+* Disks
+   * Capacity
+   * I/O
+* Interconnects (Really? Hard to measure & shouldn't be the problem most of the time...)
+
+* Software resources
+   * File descriptors?
+   * Mutex locks?
+   * Process / thread capacity?
+   * Thread pools?
+
+
+Logging
+```````
+
+Peter Bourgon has a lot of wise things to say about logging in `his brilliant article Logging v. instrumentation <https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html#:~:text=Instrumentation%20is%20for%20all%20remaining,meaningful%20number%20available%20for%20capture.>`_.
+
+* "[S]ervices should only log actionable information. That includes serious, panic-level errors that need to be consumed by humans, or structured data that needs to be consumed by machines."
+* "Logs read by humans should be sparse, ideally silent if nothing is going wrong. Logs read by machines should be well-defined, ideally with a versioned schema."
+* "A (service) never concerns itself with routing or storage of its output stream. It should not attempt to write to or manage logfiles. Instead, each running process writes its event stream, unbuffered, to stdout."
+* "Finally, understand that logging is expensive." and "Resist the urge to log any information that doesn’t meet the above criteria. As a concrete example, logging each incoming HTTP request is almost certainly a mistake."

```

Comment:
I should probably read that article carefully sometime.  I think these mostly make sense from a monitoring perspective but I also think I want a lot more information from a _debugging_ perspective.  Perhaps there are different rules that make sense for software in different stages of its lifecycle.

Nothing actionable here for now.

Reply:
Might make sense to have log levels and instruct promtail to only forward from & above "info" or "warning"  it does not overwhelm that monitoring box when we need it the most.

below info we could have as many "debug" levels as we wanted - these shouldn't be turned on in production, right?